Rails API: Ability to return error responses in json format also in development

jmbejar opened the pull request:
This PR aims to have a better experience when developing Rails app with the `--api` option.

Specifically, we found that **error pages are always being delivered in html pages in development mode**, without the chance to have errors responses rendered in json (or even xml) format. With these changes, developers would have the chance to better integrate error responses with client code in development mode. 

The approach in this PR is based on how custom Rails apps work in development mode: responses to regular http requests render html page errors and responses to xhr requests render plane text pages.
We added similar code to have error responses to Rails API requests being rendered in the requested format (json by default).

Although the primary goal was to improve the development experience, we needed to fix another problem in the way.
If you request a page with a format extension in the URL (e.g. posts/1.json) and some error happens in the middleware stack before reaching application code, you get a html response (also in production mode). However, if you include an `Accept` header with the `application/json` mime type, you get an json response including the error code and message.

In the case where format is deduced from the URL extension, the format parameter value is parsed by the router but exceptions can happen before reaching that point. For example, in Rails API applications could happen that incoming requests have a bad formatted body, causing a json parse error. This exception is raised before format is parsed by routing. This test case shows the issue: https://github.com/rails/rails/compare/rails:master...jmbejar:rails-api-json-error-response?expand=1#diff-c3f1ef70a218168e1473ce5e5245ce14R116

The solution to this secondary problem is to compute the format value from the URL using a very simple approach if it is not set yet. The format parameter value is set again once routing code executes, so it does not implies any behavior change for the application code executed after routing.

jmbejar committed:
Use URL path extension as format in bad params exception handling

jmbejar committed:
Response when error should be formatted properly in Rails API if local request

jmbejar committed:
Do not include web-console in Rails API apps

jmbejar committed:
Fix some edge cases in AD::DebugExceptions in rails api apps

jmbejar committed:
New hash syntax in AD::DebugExceptions

jmbejar committed:
Remove unneeded args in AD::DebugExceptions

jmbejar committed:
Minor cleanup in AD::DebugExceptions

edwardloveall commented:
I really like the idea of errors coming back as the format you're requesting in.

spastorino commented on actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:
```diff
@@ -65,43 +67,83 @@ def render_exception(env, exception)
       end
     end
 
+    def render_for_non_api_application(request, wrapper)
```
I'd call this `render_for_default_application`

jmbejar replied:
:+1:

spastorino commented:
Other than that seems good. Would be great if others can review ...

matthewd commented on railties/lib/rails/application/default_middleware_stack.rb:
```diff
@@ -34,7 +34,7 @@ def build_stack
           middleware.use ::Rails::Rack::Logger, config.log_tags
           middleware.use ::ActionDispatch::ShowExceptions, show_exceptions_app
-          middleware.use ::ActionDispatch::DebugExceptions, app
+          middleware.use ::ActionDispatch::DebugExceptions, app, config.api_only
```
:-1: on using `api_only` here.

I'm in two minds as to whether we should just always give a JSON error for a request that wants JSON... but if we want to keep the current HTML-for-JSON behaviour for "regular" applications, this should be based on a new config setting (which would get set to true by the API generator).

If "make Rails better for APIs" means "introduce some new configurable features, and some new defaults", we'll all win. But if it means "introduce an expanding number of places that are directly checking the value of `config.api_only`", I for one will be very sad.

spastorino replied:
@matthewd agreed about the new config. That could also be done for the session as well, etc.

jmbejar replied:
@matthewd @spastorino I like the idea of having a separate config. Will do.

jmbejar replied:
@matthewd @spastorino I've refactored this to have a separate config option. You can see the related changes here: https://github.com/rails/rails/compare/ffa2a5de6414030a66ca2f61ca2ab725096b4158...jmbejar:rails-api-json-error-response

I initially tried having a boolean flag but it seems to fit better if we have a config setting with symbol values that describe the behavior of DebugExceptions. For now, we have `:default` and `:api` modes.

One drawback of this approach where we have separate config settings: if some app uses `api_only = false` and `debug_exception_response_format = :api` for some reason, it will end up having JSON responses for error requests while the app behaves as a default application.

This is a point of concern for me, but I understand that it would be caused by developer decisions since they will have more flexibility to define the behavior. In any case, I'm still pushing these commits to the PR to have more feedback and discussion.

jmbejar committed:
Improve regexp in AC::Http::Parameters

jmbejar committed:
Better name for method in DebugExceptions middleware

jmbejar committed:
Add debug_exception_response_format config to configure DebugException
behavior

jmbejar committed:
DebugException initialize with a response_format value

jmbejar committed:
Mention the debug_exception_response_format config in guides

jmbejar committed:
Fix indent in generated Rails API env file

jmbejar committed:
debug_exception_response_format needs to be writeable in Configuration

jmbejar committed:
Update Changelog with the added response_format option in AD::DebugExceptions

jmbejar commented:
Just to make sure we're all in the same page, we need to wait for rails/web-console#152 (almost ready!) before going forward with this PR.

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/Gemfile:
```diff
@@ -36,12 +36,21 @@ group :development, :test do
+<%- if options.api? -%>
+  # Access an IRB console on exception pages
+  <%- if options.dev? || options.edge? -%>
+  # gem 'web-console', github: 'rails/web-console'
```
Why we are even adding web-console commented? If the app is API only we will not care about web-console

jmbejar replied:
Since we decided to have a separate config value (`debug_exception_response_format`) to setup the debug response format, it's still valid to use `web-console` in an API only app. You can configure `debug_exception_response_format = :default` and `api_only = true`, so you should get HTML error pages on development where the console can be used.

In the other hand, JSON responses are the default configuration (`debug_exception_response_format = :api`), and for this reason it makes sense to do not include `web-console` by default. However, we're leaving it commented because it's still supported.

In a related note, we were waiting for changes in `web-console` to fully support this case (https://github.com/rails/web-console/pull/152), but that's already merged :)

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/config/environments/development.rb.tt:
```diff
@@ -12,6 +12,15 @@ Rails.application.configure do
+
+  # Return error responses in the format requested by the client
+  # or default to JSON format.
+  # This option is useful for Rails API only applications developers
```
We can remove this second paragraph

jmbejar replied:
Agreed :+1:

jmbejar committed:
Adjust comment in development.rb template file for app generator

rafaelfranca commented:
You need to rebase this branch too

jmbejar committed:
Do not add format key to request_params

I did this change but it is affecting how the request params end up
after being processed by the router.

To be in the safe side, I just take the format from the extension in the
URL when is not present in those params and it's being used only for the
`Request#formats` method

jmbejar committed:
Avoid warning because of the mime type

jmbejar committed:
Add tests for api_only configuration setting

jmbejar committed:
Make debug_exception_response_format config depends on api_only when is not set

spastorino commented on railties/lib/rails/generators/rails/app/templates/config/environments/development.rb.tt:
```diff
@@ -23,7 +23,12 @@ Rails.application.configure do
+  # Return error responses in the format requested by the client
+  # or default to JSON format.
+  config.debug_exception_response_format = :api
+  <%- end -%>
```
You can remove this now :)

spastorino commented on railties/test/generators/api_app_generator_test.rb:
```diff
@@ -40,12 +40,18 @@ def test_api_modified_files
       assert_no_match(/gem 'web-console'/, content)
       assert_match(/gem 'active_model_serializers'/, content)
+
+      assert_no_match(/gem 'web-console'/, content)
```
This is duplicating line 41

spastorino commented on railties/test/generators/api_app_generator_test.rb:
```diff
@@ -40,12 +40,18 @@ def test_api_modified_files
 
+    assert_file "config/environments/development.rb" do |content|
+      assert_match(/config.debug_exception_response_format = :api/, content)
+    end
```
Also remove this

spastorino commented on actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:
```diff
@@ -17,6 +17,7 @@ def debug_params(params)
         clean_params = params.clone
         clean_params.delete("action")
         clean_params.delete("controller")
+        clean_params.delete("format")
```
Remove this

jmbejar committed:
We don't need to set config.debug_exception_response_format given that :api is the default value for only API apps

jmbejar committed:
Remove duplicated assertion from api_app_generator tests

spastorino commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -67,6 +67,8 @@ def formats
           v = if params_readable
             Array(Mime[parameters[:format]])
+          elsif format_from_path_extension
+            [Mime[format_from_path_extension]]
```
You're calling this method twice without memoizing the result. I'd store the result in a local var or try to memoize in some way.

jmbejar committed:
Avoid calling AD::MimeNegotiation#format_from_path_extension method twice
